Shell: Add duplicate route validation for sibling elements

SubhikshaSf4851 opened the pull request:
<!-- Please let the below note in for people that find this PR -->
> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could [test the resulting artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from this PR and let us know in a comment if this change resolves your issue. Thank you!

### Root Cause

The Shell routing system allowed duplicate routes to be registered for sibling elements (e.g., two ShellContent items in the same ShellSection with route "MapPage"). This caused confusing navigation behavior where `OnNavigatedTo` would not be called when navigating between pages with duplicate routes.

The user discovered this was the root cause of issue #14000 and requested that MAUI throw an exception when duplicate routes are detected, to prevent this confusing scenario.

### Description of Change

Added **sibling-based duplicate route validation**:

1. **`Routing.ValidateForDuplicates(Element, string)`** - New validation method that:
   - Checks if a route is already assigned to a sibling element (elements with same parent)
   - Allows same route at different hierarchy levels (e.g., `//one/content` and `//two/content`)
   - Throws `ArgumentException` with clear error message when duplicate found

2. **`BaseShellItem.Route` property setter** - Now calls validation before setting route

3. **5 comprehensive unit tests** - Verify:
   - Duplicate sibling routes throw exception
   - Same route in different parents is allowed (hierarchical routing)
   - Changing/removing routes frees them for reuse
   - Reassigning same route to same element is idempotent

4. **Cleanup** - Removed duplicate `AddTopTab()` call in Issue6878 test that would now throw

**This is a philosophy change:**
- **Before:** Duplicate routes silently accepted, caused confusing behavior
- **After:** Duplicate sibling routes rejected immediately with clear error message

### Key Technical Details

**Sibling-based validation:**
- Only checks duplicates among elements with the same parent (`element.Parent.LogicalChildrenInternal`)
- Respects Shell's hierarchical routing structure
- No global state needed

**Why sibling-based, not global?**
- Shell supports hierarchical routing where same route can exist at different levels
- Example: `//route1/content` and `//route2/content` are both valid
- Global tracking would incorrectly reject this valid scenario

### What NOT to Do (for future agents)

- ❌ **Don't use global route tracking (HashSet<string>)** - Breaks Shell's hierarchical routing. Reviewers rejected this approach because same route at different hierarchy levels is valid.
- ❌ **Don't use ConcurrentDictionary or shared state** - Not thread-safe and causes lifecycle issues (routes never removed when elements deleted)
- ❌ **Don't validate at Shell level** - Validation must be at BaseShellItem.Route setter for immediate feedback

### Issues Fixed

Fixes #14000

### Tested the behavior in the following platforms

- [x] Windows
- [x] Android
- [x] iOS
- [x] Mac

SubhikshaSf4851 committed:
Fix for throw exception for duplicate items

SubhikshaSf4851 committed:
Updated namings

SubhikshaSf4851 committed:
updated Fix in common place

SubhikshaSf4851 committed:
Updated Test Sample

SubhikshaSf4851 committed:
updated unique route in test sample

SubhikshaSf4851 committed:
Updated Fix

SubhikshaSf4851 committed:
Updated copilot suggestion

SubhikshaSf4851 committed:
Updated Exception details

StephaneDelcroix committed:
Move duplicate routes test from DeviceTests to Core.UnitTests

The test ShouldThrowsArgumentExceptionOnDuplicateRoutes doesn't require
any platform handlers - it's testing pure exception throwing logic.
This should be a unit test, not a device test.

Moved to: src/Controls/tests/Core.UnitTests/ShellTests.cs
Removed from: src/Controls/tests/DeviceTests/Elements/Shell/ShellTests.cs

StephaneDelcroix reviewed (changes requested):
this should be a unit test
the linked issue looks unrelated

SubhikshaSf4851 commented:
@StephaneDelcroix The fix were made for the user comment [here](https://github.com/dotnet/maui/issues/14000#issuecomment-1474008920)

StephaneDelcroix committed:
Fix duplicate route detection to use sibling-based validation

- Changed from global route tracking to sibling-based duplicate detection
- Routes are now only checked for duplicates among siblings (same parent)
- This allows the same route name in different parts of the Shell hierarchy
- Removed unused ConcurrentDictionary and global route tracking
- Reverted device test changes that were workarounds for global tracking
- Added comprehensive unit tests for sibling-based validation:
  - DuplicateSiblingRoutesShouldThrowArgumentException
  - SameRouteInDifferentParentsIsAllowed
  - ChangingRouteAllowsReuseAmongSiblings
  - RemovingElementClearsRoute
  - ReassigningSameRouteToSameElementDoesNotThrow

StephaneDelcroix commented:
## Changes Made to Fix Implementation

I've updated this PR to use **sibling-based duplicate route detection** instead of global tracking. Here's a summary of the changes:

### Why the Change Was Needed

The original implementation used a global `ConcurrentDictionary` to track all routes, which broke legitimate Shell navigation patterns. Shell routes are hierarchical - the same route name (like "content") can exist at different levels of the hierarchy:

```
//flyout1/section1/content   ← "content" here
//flyout2/section2/content   ←
```

Global tracking would incorrectly flag this as a duplicate.

### Implementation Summary

**`Routing.cs`** - New `ValidateForDuplicates()` method that:
- Only checks for duplicates among sibling elements (same parent)
- Skips validation for non-user-defined routes (auto-generated ones)
- Returns early if reassigning the same route to the same element

**`BaseShellItem.cs`** - Route property setter calls `ValidateForDuplicates()` before setting the value

### Test Coverage

Added 5 unit tests in `ShellTests.cs`:
1. `DuplicateSiblingRoutesShouldThrowArgumentException` - Verifies exception for duplicate sibling routes
2. `SameRouteInDifferentParentsIsAllowed` - Confirms hierarchical routes work correctly
3. `ChangingRouteAllowsReuseAmongSiblings` - Route can be reassigned after changing
4. `RemovingElementClearsRoute` - Removed elements free up their routes
5. `ReassigningSameRouteToSameElementDoesNotThrow` - Idempotent route assignment

All 85 Shell unit tests pass with this implementation.

### Note on Issue Reference

The PR description references #14000, but that issue is about "OnNavigatedTo not being called" which appears unrelated to duplicate route detection. You may want to clarify which issue this PR actually addresses, or create a new issue specifically for duplicate route detection if one doesn't exist.

StephaneDelcroix reviewed (dismissed):
✅ **Approved**

The sibling-based duplicate route detection implementation is correct and well-tested. 

**Key points:**
- Properly respects Shell's hierarchical navigation model
- Only prevents duplicate routes among siblings (same parent)
- Allows the same route name at different hierarchy levels (e.g., `//one/content` and `//two/content`)
- All 85 Shell unit tests pass
- 5 new comprehensive unit tests cover the feature

The implementation is minimal, focused, and doesn't break existing functionality.

rmarinho commented:
/azp run maui-pr-devicetests

SubhikshaSf4851 committed:
Removed duplicate AddTopTab call in Issue6878 shell test case
